File: apps/mobile/lib/core/format/money.dart
import 'package:intl/intl.dart';

/// Price formatting. All amounts travel as integer minor units (cents) with
/// an explicit ISO-style currency code, exactly as the API sends them.
///
/// Whole amounts drop the `.00` (a latte is "25 MDL", not "25.00 MDL");
/// fractional ones keep two digits. Symbol placement follows the locale:
/// English puts a sign in front ("$4.50"), Russian after ("4,50 $"), and
/// currencies without a widely known sign are written as a code after the
/// number in both.
abstract final class Money {
  static const _prefixSigns = {'USD': r'$', 'EUR': '€', 'GBP': '£', 'THB': '฿', 'IDR': 'Rp'};

  /// Codes spelled out after the number. The Transnistrian rouble has no ISO
  /// code; locally it is written "руб.".
  static const _suffixCodes = {'AED': 'AED', 'MDL': 'MDL', 'RUP': 'RUP'};
  static const _suffixCodesRu = {'RUP': 'руб.'};

  static String format(int cents, String currency, {String locale = 'en'}) {
    final negative = cents < 0;
    final abs = cents.abs();
    final digits = abs % 100 == 0 ? 0 : 2;
    final number = NumberFormat.decimalPatternDigits(locale: locale, decimalDigits: digits).format(abs / 100);
    final isRussian = locale.toLowerCase().startsWith('ru');
    final code = currency.toUpperCase();

    final String body;
    final prefix = _prefixSigns[code];
    if (prefix != null && !isRussian) {
      body = '$prefix$number';
    } else if (prefix != null) {
      body = '$number $prefix';
    } else {
      final label = (isRussian ? _suffixCodesRu[code] : null) ?? _suffixCodes[code] ?? code;
      body = '$number $label';
    }
    return negative ? '−$body' : body;
  }
}

File: libs/utils/src/lib/money.ts
/**
 * Prices for people to read. Amounts travel as integer minor units (cents)
 * with a currency code, exactly as the API sends them; the rules match the
 * Flutter app's `Money.format`, so a latte costs the same "38 MDL" on every
 * screen.
 *
 * - Whole amounts drop the cents ("38 MDL", not "38.00 MDL"); others keep
 *   two digits with the language's own separators ("4,50 MDL" in Russian).
 * - Currencies with a well-known sign put it before the number in English
 *   ("$4.50") and after it in Russian ("4,50 $").
 * - The rest are written as a code after the number. The Transnistrian
 *   rouble has no ISO code; in Russian it is «руб.», as it is written there.
 * - Negative amounts start with a real minus sign ("−5 MDL").
 */

import type { Currency } from '@takeaway/shared-types';

const PREFIX_SIGNS: Readonly<Record<string, string>> = { USD: '$', EUR: '€', GBP: '£', THB: '฿', IDR: 'Rp' };

const RUSSIAN_CODES: Readonly<Record<string, string>> = { RUP: 'руб.' };

// A no-break space keeps "38 MDL" on one line in a narrow card.
const NBSP = '\u00a0';
const MINUS = '\u2212';

export interface FormatMoneyOptions {
  /** Round to whole units — for KPI cards and chart labels, where cents are noise. */
  round?: boolean;
}

export function formatMoney(
  cents: number,
  currency: Currency | string | null | undefined,
  lang = 'ru',
  options: FormatMoneyOptions = {},
): string {
  const minor = options.round ? Math.round(cents / 100) * 100 : Math.round(cents);
  const abs = Math.abs(minor);
  const number = numberFormat(lang, abs % 100 === 0 ? 0 : 2).format(abs / 100);
  const russian = isRussian(lang);
  const code = (currency ?? '').toUpperCase();

  let body: string;
  if (!code) {
    // A total across brands with different currencies has no single unit.
    body = number;
  } else if (PREFIX_SIGNS[code]) {
    body = russian ? `${number}${NBSP}${PREFIX_SIGNS[code]}` : `${PREFIX_SIGNS[code]}${number}`;
  } else {
    body = `${number}${NBSP}${(russian && RUSSIAN_CODES[code]) || code}`;
  }
  return minor < 0 ? `${MINUS}${body}` : body;
}

/** Whether a language tag is Russian ("ru", "ru-RU", "RU"). */
export function isRussian(lang: string | null | undefined): boolean {
  return (lang ?? '').toLowerCase().startsWith('ru');
}

const numberFormats = new Map<string, Intl.NumberFormat>();

function numberFormat(lang: string, digits: number): Intl.NumberFormat {
  const key = `${lang}|${digits}`;
  let format = numberFormats.get(key);
  if (!format) {
    const options = { minimumFractionDigits: digits, maximumFractionDigits: digits };
    try {
      format = new Intl.NumberFormat(lang || 'ru', options);
    } catch {
      // A malformed language tag throws; the amount still has to show.
      format = new Intl.NumberFormat('ru', options);
    }
    numberFormats.set(key, format);
  }
  return format;
}
